### PR TITLE
Allow server side occlusion culling.

### PR DESCRIPTION
**Pull request by lhofhansl**
1. Moved occlusion culling code from ClientMap to Map.

2. Changed transparency condition from 
```
		const ContentFeatures &f = nodemgr->get(n);
		if(f.solidness == 0)
			is_transparent = (f.visual_solidness != 2);
		else
			is_transparent = (f.solidness != 2);
```
to
`f.drawtype == NDT_NORMAL`
This is needed so the same code can be used on client and server without virtual dispatching in hot code path.
(see ContentFeature::updateTextures for why this is correct)

3. Optionally enable occlusion code on the server, using the player eye-position as reference point.

In my tests this saved between 50 and 80% of the blocks sent to the server when block_send_optimize_distance is set to the maximum with no visual impact.

### CODE REVIEW

**Comment by Fixer-007** [comment]
What about server/network lags?

**Comment by lhofhansl** [comment]
@Fixer-007 what about it? :)
This will generally send less data, but it can't do anything about a laggy network.

**Comment by Fixer-007** [comment]
If network lags and I walking on server, will culling then also lag?

File: src/clientiface.cpp [hunk 1]
```diff
@@ -298,6 +301,12 @@ void RemoteClient::GetNextBlocks (
 					if(block->getDayNightDiff() == false)
 						continue;
 				}
+
+				if (occ_cull && !block_is_invalid &&
+				    env->getMap().isBlockOccluded(block, cam_pos_nodes))
```
Comment (SmallJoker) [inline src/clientiface.cpp]:
```C++
if (occ_cull && !block_is_invalid &&
		env->getMap().isBlockOccluded(block, cam_pos_nodes)) {
```

File: src/clientiface.cpp [hunk 2]
```diff
@@ -197,6 +197,9 @@ void RemoteClient::GetNextBlocks (
 	s32 nearest_sent_d = -1;
 	//bool queue_is_full = false;
 
+	v3s16 cam_pos_nodes = floatToInt(camera_pos, BS);
```
Comment (SmallJoker) [inline src/clientiface.cpp]: `const`?

**Comment by SmallJoker** [comment]
@lhofhansl  Is this considered as an alternative method to prevent blocks from being sent to the clients for pull #5309 ?

**Comment by lhofhansl** [comment]
@Fixer-007 The server side culling is independent on that.

**Comment by lhofhansl** [comment]
@SmallJoker This is definitely related to #5309 and #4811 . 
Right now I'd suggest to keep all of them. Upon more testing we can retire one or more of the other.

**Comment by lhofhansl** [comment]
But in fact this can replace all the other server side optimizations.

**Comment by Fixer-007** [comment]
@lhofhansl Come to VanessaE Dreambuilder Survival or Creative (or other VanessaE's servers), they have this patch.

**Comment by lhofhansl** [comment]
@Fixer-007 Are you saying something broke or renders incorrectly or is slowing down? I tested with my own server and mostly focused in the number of blocks the client reports as "seen".

File: src/clientiface.cpp (hunk 2 above)
Comment (lhofhansl) [inline src/clientiface.cpp]: I knew that would come up. :)
We're kinda schizophrenic about this. I'll change it.

File: src/clientiface.cpp (hunk 1 above)
Comment (lhofhansl) [inline src/clientiface.cpp]: Oops. That one slipped through.

**Comment by Fixer-007** [comment]
No, just for test. I've walked on it a bit and have not noticed anything suspicious yet on VE-S.

**Comment by lhofhansl** [comment]
@Fixer-007 Cool. I might do that. Never actually player on a public server :)

What this does it just has the server filter the blocks to send by what the play would actually see. There are other mechanisms that take the player's position into account, so I would not expect this to cause new visual lag.

I'll do some tests with some scenes and compare the day-night spread optimization (which leads to rendering issues) to the occlusion culling.

**Comment by lhofhansl** [comment]
Update style and const declaration. No functional changes.

**Comment by lhofhansl** [comment]
To make review easier...
1. just moved the occlusion code from ClientMap to Map
2. isOccludes is now a member, removed the Map and iNodeDef parameter accordingly
3. Move the occlusion logic in ClientMap.updateDrawlist into its own method
4. Change the transparent logic as indicated in the initial comment (it really does not change when a node is considered transparent)
5. Called the isBlockOccluded also ClientIface.GetNextBlocks.

The change looks bigger than it is, because I had to move some code around.

**Comment by lhofhansl** [comment]
Looks like max_block_send_distance is set only to 3 on that server (I can see the edge of the world with a v_range > 50). Is that true?

**Comment by lhofhansl** [comment]
Some numbers: 

Water scene:

-|Default |#5309 (no underground opt)|#4811 (opt distance set to max)|#5320 (this PR)
-|-|-|-|-
Blocks Rendered | 331  | 508 | 506 | 506
Blocks in view | 426 | 1771 | 2954 | 1098

Land scene:

-|Default |#5309 (no underground opt)|#4811 (opt distance set to max)|#5320 (this PR)
-|-|-|-|-
Blocks Rendered | 250  | 345 | 341 | 341
Blocks in view | 448 | 1506 | 2043 | 1045

v_range was 200

screenshot_20170228_210452

screenshot_20170228_212229

**Comment by lhofhansl** [comment]
So this is clearly superior to #5309 , and this will always render correctly (if the culling is incorrect it is incorrect in exactly the same way the client culling is incorrect).

**Comment by lhofhansl** [comment]
Only concern is potential CPU spent on the server to do occlusion culling for each player. The upside is that at any point only blocks are sent to the client that the player can actually see.

**Comment by Fixer-007** [comment]
So amount of rendered stuff will increase by 35-50% ? Thats a lot of fps drain :(

**Comment by lhofhansl** [comment]
@Fixer-007 We'll keep the day-night optimization - which has the rendering errors. Server side occlusion help even on top of that. But when you want correct rendering, than occlusion culling can give you that and save blocks (in tight caves that can be saving 90% or more).

**Comment by lhofhansl** [comment]
>
>    +# If enabled the server will perform map block occlusion culling based on
>    +# on the eye position of the player.
>
>Is occlusion culling also disabled on the client? I guess it is because it is less effective so would not be >of any use.

No it's not. This just makes sure that the server does not send blocks the player cannot see. The since the player moves around, there will be cached blocks on the client that still should not be rendered, so client side culling is still useful.

>
>    +# The client will not longer receive most invisible
>    +# so that the utility of noclip mode is reduced.
>
>What does this mean? Can we still fly underground with noclip? (essential for dev work).

Flying around with noclip still works, but the client will naturally have fewer blocks to see, that's all.

>
>What is it about server occlusion culling that makes it more effective than on the client?

It doesn't. It's the same logic. It just happens *before* the blocks are sent to the client, rather than after.

**Comment by paramat** [comment]
I see, thanks, seems good.

**Comment by Fixer-007** [comment]
I played with this in singleplayer and have not noticed anything suspicious yet.

**Comment by lhofhansl** [comment]
Just to make 100% sure, you set server_side_occlusion_culling to true in your minetest.conf file, right? :)

**Comment by lhofhansl** [comment]
I now tried in nether. As well as in small caves.
In small caves it's loading 46 blocks instead of almost 1400 blocks (with #4811 optimize distance set to maximum). Yes, 46 instead of 1400 - occlusion culling is most effective in caves.

Rendering is correct in all cases.

**Comment by Zeno-** [comment]
Sending less data from MP servers == :+1:

**Comment by Fixer-007** [comment]
>server_side_occlusion_culling to true in your minetest.conf file, right? :)

Ugh, for real? Who needs to write? server owner or client or both?

**Comment by lhofhansl** [comment]
Heh. Glad I asked. On the server only.
I should mage like make it the default anyway. In no testing have I seen the slightest problem.

**Comment by lhofhansl** [comment]
Let me turn this default on.

**Comment by lhofhansl** [comment]
Turned on by default now.

**Comment by lhofhansl** [comment]
As discussed in #5309 perhaps now it's time to increase `block_send_optimize_distance` to 6

**Comment by lhofhansl** [comment]
As discussed in #5309 perhaps now it's time to increase `block_send_optimize_distance` to 6

**Comment by lhofhansl** [comment]
Some typical scenes:
-|default|distance = 6|distance = 6 + this PR
-|-|-|-
land1|396|578|412
land2|549|700|635
water|426|644|560
cave|444|704|45

I think it's safe to turn this on and set `block_send_optimize_distance` to 6

**Comment by lhofhansl** [comment]
Anything else I need to do with this?

This saves a bunch of block sending from server to client. Since this uses the same algorithm on the server that we have been using on the client this will never miss blocks that client occlusion culling would not filter out as well.

**Comment by paramat** [comment]
Sorry for quiet dev, many devs seem to be absent or busy at the moment, and the few left overwhelmed, i think this has a lot of potential, we are very interested and i'll try to look as soon as i can.
Are you back from travelling yet? Hopefully in a while we can add you as a core dev, if you are ok with that, we certainly need the help.

**Comment by lhofhansl** [comment]
(Yeah back from travel - work still keeping me busy, though. Happy to help wherever I can, with code, reviewing, keeping the code sane, etc)

**Comment by lhofhansl** [comment]
Oh and FPS should be unaffected. The complexity of the rendered scene is exactly the same.
Perhaps slightly better jitter because fewer unnecessary blocks are loaded into the client.

**Comment by Zeno-** [comment]
👍

**Comment by rubenwardy** [comment]
It's worth benchmarking the server, and seeing if/how much it reduces load

**Comment by lhofhansl** [comment]
Ran it through the Linux perf counters. The difference I'm seeing are within the noise. Without this PR I see perhaps a little more time spent in malloc (1% or so). With the PR I see (also 1% or so) in Map::getNodeNoEx and Map::getSectorNoGenerateNoExNoLock. All within the noise of a few runs.

Like the day-night light test optimization to see whether a block is near ground level, the blocks needs to be actually loaded in order to do occlusion culling (again just like for the day-night optimization), so I would not expect lighter load on the server; just lighter load on the network.

(Next step is making the occlusion more effective - i.e. re-visiting #4933 )

**Comment by lhofhansl** [comment]
Put some counters into the server to verify that the client's "blocks in range" does reflect the number of blocks sent from the server. It does :)

**Comment by HybridDog** [comment]
This also helps against cheating, doesn't it?
If some player e.g. changes the source code to get noclip and sets the drawtype of nyancat rainbows to glasslike in nodedef.cpp, he/she can not longer fly around underground and see all the nyancats stuck in the stone.
And local_map_saving does not longer store redundant blocks, which results in a smaller map.sqlite.

**Comment by lhofhansl** [comment]
@HybridDog yes it would incidentally help with that, but you couldn't rely on that. You can just fly around and look at the scene from different angles and over the time the client would cache the scene from all sides and you'd be able to do the very same hacks.

**Comment by HybridDog** [comment]
Can you automatically temporary disable server side occlusion culling for players which aren't in caves (test if they stay in sunlight) when the server has a high CPU usage?

**Comment by paramat** [comment]
^ That's not a good idea.

**Comment by lhofhansl** [comment]
It would only be a good idea with Occlusion Culling would be computationally expensive.
It's not and the blocks it'd load need to loaded anyway. Extra CPU is very light (within the measuring noise)

**Comment by sorcerykid** [comment]
How computationally expensive is it on a high-traffic server? At peak hours I average about 45-50 connected players. The potential gains on the network side are appealing, but I'm concerned this patch might prove to be a performance drain for the CPU. Thanks.

**Comment by lhofhansl** [comment]
By far the main drain is handling ABMs (also done per player). Occlusion culling is not measurable compared to the ABM cost, it's within the noise of the measurement.
Loading and serving all these blocks is also more expensive than the extra check for occlusion culling (which is done only on blocks that passed range and the day-night-light checks), and block loading also needs to be done per player.

In the end I have no way to experimentally verify that with dozens of players, though.

If it turns out to be a CPU drain for a specific server it can always be turned off via the config option - but I really doubt it will ever be a problem.

**Comment by sofar** [comment]
:+1:, this will be good for mp servers

**Pull request merged**